aiosmtpd/smtp.py

```diff
@@ -505,29 +503,34 @@ def smtp_DATA(self, arg):
             return
         yield from self.push('354 End data with <CR><LF>.<CR><LF>')
         data = []
-        self.num_bytes = 0
+        num_bytes = 0
+        size_exceeded = False
         while not self.connection_closed:
             line = yield from self._reader.readline()
             if line == b'.\r\n':
+                if data:
+                    data[-1] = data[-1].rstrip(b'\r\n')
                 break
-            self.num_bytes += len(line)
-            if self.data_size_limit and self.num_bytes > self.data_size_limit:
+            num_bytes += len(line)
+            if (not size_exceeded) and self.data_size_limit and \
+                    num_bytes > self.data_size_limit:
```

Comment:
A few style nits here.  I'd prefer something like this:

```
            if (not size_exceeded and
                    self.data_size_limit > 0 and
                    num_bytes > self.data_size_limit):
```

One thing I noticed is that we sometimes use the "false-y-ness" of `self.data_size_limit` but I think the attribute must always be an integer.  If so, I prefer to compare it against `> 0`.  However, if it can also be `None` then a false-y conditional is okay.  I'm not sure that we document it, but seeing other tests in the code makes me think it has to be an integer.  What do you think?